marlowe/specification/ReadMe.md

```diff
@@ -0,0 +1,346 @@
+```
+In the above, the `BuiltinByteString` is the hash of the serialized continuation of the contract. If `MerkleizedTxInput` is supplied in a redeemer, then the `ScriptContext` for the transaction must also contain an extra entry in its `txInfoData . scriptContextTxInfo` map from `DatumHash` to `Datum` for the serialized continuation of the contract.
+
+
+### Relationship between Marlowe Validator and Semantics
+
+The arguments are `computeTransaction` must be constructed as follows:
+1. The `txInterval` of `TransactionInput` is derived from the `txInfoValidTimeRange . scriptContextInfo` of the `ScriptContext`, as detailed below.
+2. The `txInputs` of `TransactionInput` is derived from the `MarloweInput` provided as the `Redeemer` and the `txInfoData . scriptContextTxInfo` of the `ScriptContext`, as detailed below.
+3. The `State` is the `marloweState` of the `MarloweData` provided as the `Datum`.
+4. The `Contract` is the `marloweContract` of the `MarloweData` provided as the `Datum`.
+5. The new `Datum` at the script address is the `MarloweData` with the same `marloweParams` as originally, but with the new `txOutState` and `txOutContract` of the `TransactionOutput`.
+
+![Relationship between Marlowe validator and semantics.](semantics2plutus.svg)
+
+
+### Specification
+
+Consider the application of the Marlowe validator and Marlowe semantics:
+```haskell
+validationResult = smallMarloweValidator marloweData marloweInput scriptContext
+
+transactionOutput = computeTransaction transactionInput inState inContract
+```
+
+Furthermore, let `marloweValidatorHash :: ValidatorHash` be the hash of the Marlowe semantics script.
+
+The validation fails (via returning `False` for `validationResult` or via the throwing of an error) if any of the following constraints does not hold.
+
+
+#### *Constraint 1.* Typed validation
+
+The datum, redeemer, and script context deserialize to the correct types.
+
+Let `datum :: BuiltinData` and `redeemer :: BuiltinData` be the datum and redeemer in the script witness for spending the Marlowe UTxO, and let `scriptContext' :: BuiltinData` be the script context data.
+```haskell
+fromBuiltinData datum          ≡ (Just marloweData   :: Maybe MarloweData)
+fromBuiltinData redeemer       ≡ (Just markloweInput :: Maybe MarloweInput)
+fromBuiltinData scriptContext' ≡ (Just scriptContext :: Maybe ScriptContext)
+```
+
+
+#### *Constraint 2.* Single Marlowe script input
+
+The output of exactly one Marlowe UTxO is spent in the transaction.
+```haskell
+filter (isMaroweAddress . txOutAddress . txInfoResolved) (txInfoInputs $ scriptContextTxInfo scriptContext) ≡ [TxInInfo _ (TxOut _ inValue _)]
+  where isMarloweAddress (Address (ScriptCredential hash)) = hash == marloweValidatorHash
+        isMarloweAddress _ = False
+```
+Hereafter we use `inValue` as the script value being spent.
+
+
+#### *Constraint 3.* Single Marlowe output
+
+There is a single output to the Marlowe script address unless the contract closes.
```

Comment:
in which case there is no output.

Reply:
fixed in 8be04323e4ec24c4e923ed214094e85c916ab8e0